### ISSUE TITLE
When invoking external command inside custom command, how to correctly link the stdout/stderr so the external command can react correctly.

### ISSUE THREAD

**Issue by edwinjhlee** [nushell/nushell#13156]
### Question

```bash
def --wrapped f0 [...args] { bash -c 'ls'; }

def --wrapped f1 [...args] { bash -c 'ls'; echo "do something after the command" ; }
```

In f0, the `bash -c 'ls'` stdout and stderr is linked to f0.
However, the f1 `bash -c 'ls'` is not.

So when I invoke `f1 | cat`, the `bash -c 'ls'` cannot correctly react as expected ( the stdout is a pipe )

How can we invoke `bash -c 'ls'` so the stdout and stderr can link to the f1's stdout and stderr 

### Additional context and details

I am providing support on nushell invoking x-cmd.
Here is a milestone demo: 

https://www.x-cmd.com/mod/nu

Currently, I cannot make the following line to react correctly when the `___x_cmd_nu_rc_xbinexp`'s output is a pipe.

https://github.com/x-cmd/x-cmd/blob/33a0346e2a80d04dde0eb2e7a4342ba14d836ff8/mod/nu/lib/rc/rc.nu#L34
```
bash $"($env.HOME)/.x-cmd.root/bin/xbinexp" ...$args
```

Thank you.

**Comment by edwinjhlee** [comment]
BTW, can x-cmd be added to the list below ? 

https://github.com/nushell/nushell?tab=readme-ov-file#officially-supported-by

**Comment by fdncred** [comment]
I'm trying to understand what you're doing but missing the point.

It looks like you're mixing internal and external commands with `f1`. What do you expect the output to look like with `f1 | cat`?

> BTW, can x-cmd be added to the list below ?

Probably, but it's more likely to be added to our awesome-nu repo.

**Comment by fdncred** [comment]
is this what you want?
```nushell
❯ def --wrapped f1 [...args] { print (bash -c 'ls'); echo "do something after the command" ; }
❯ f1 | cat
CITATION.cff
CODE_OF_CONDUCT.md
CONTRIBUTING.md
Cargo.lock
Cargo.toml
Cross.toml
LICENSE
README.md
assets
benches
crates
devdocs
docker
rust-toolchain.toml
scripts
src
test
tests
toolkit.nu
typos.toml
wix
do something after the command
```

**Comment by edwinjhlee** [comment]
> is this what you want?
> 
> ```nushell
> ❯ def --wrapped f1 [...args] { print (bash -c 'ls'); echo "do something after the command" ; }
> ❯ f1 | cat
> CITATION.cff
> CODE_OF_CONDUCT.md
> CONTRIBUTING.md
> Cargo.lock
> Cargo.toml
> Cross.toml
> LICENSE
> README.md
> assets
> benches
> crates
> devdocs
> docker
> rust-toolchain.toml
> scripts
> src
> test
> tests
> toolkit.nu
> typos.toml
> wix
> do something after the command
> ```

YES.
I can see the solution is working. For it converts `bash -c 'ls'` result as a string.

But it is not I expect.

```
def --wrapped f1 [...args] { print (bash -c 'ls'); echo "do something after the command" ; }
```

This solution will make 'bash -c 'ls'' stdout as a pipe, not a tty.


```
def --wrapped f0 [...args] { bash -c 'ls'; }

def --wrapped f1 [...args] { bash -c 'ls'; echo "do something after the command" ; }
```

In the f0:

```
# the bash -c 'ls' can detect itself's stdout is tty
f0  

# the bash -c 'ls' can detect itself's stdout is pipe, so ls will print one item per line.
f0 | cat 
```

The `bash -c "ls"` behavior can change according to the f0


This solution will make `bash -c "ls"` always behave in the pipe stdout mode.

```
def --wrapped f1 [...args] { print (bash -c 'ls'); echo "do something after the command" ; }
```

**Comment by IanManske** [comment]
Currently, I don't think there's a way to do this in Nushell, but you can use a subshell to accomplish what you want:
```nushell
def --wrapped f1 [] { nu -c 'bash -c ls; print "do something after the command"' }
```
Here, `ls` will detect if stdout is the terminal or a pipe and change its output accordingly.

**Comment by edwinjhlee** [comment]
That is a pity. I need to set the environment variables in the nullshell after the command is finished.

```nushell

def --wrapped f1 [] { 

   bash xxx.sh      # The command executed.
   
   # setting the environment variables according to the filesystem changed by the  'bash xxx.sh'
}
```

I have read the hooks introduction in the nushell website. 
Is there a kind of hook, like trap EXIT or trap RETURN in bash, so I can do something before the function exits ?


Thank you. @fdncred @IanManske 






> Currently, I don't think there's a way to do this in Nushell, but you can use a subshell to accomplish what you want:
> 
> ```nushell
> def --wrapped f1 [] { nu -c 'bash -c ls; print "do something after the command"' }
> ```
> 
> Here, `ls` will detect if stdout is the terminal or a pipe and change its output accordingly.

**Comment by devyn** [comment]
> Is there a kind of hook, like trap EXIT or trap RETURN in bash, so I can do something before the function exits ?

No, most of our hooks have to do with prompt and REPL lifecycle, there isn't anything for function lifecycle in there.

I think it's kind of a fundamental challenge for us, because we can't allow you to change environment variables at the end of a stream. The environment is only mutable while the command has control, but when the command returns a stream it no longer has control.

So you can either collect the output into a string in memory and then do what you need to do with `$env`, or you can return it as a stream but you won't be able to affect `$env` anymore at the end.

**Comment by WindSoilder** [comment]
Closing the issue because the question is answered.
